### ISSUE TITLE
[SR-15802] Duplicate symbols with multi-arch swiftpm builds and xcframeworks

### ISSUE THREAD

**Issue by keith** [apple/swift-package-manager#5733]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15802      |
   |Radar             | None         |
   |Original Reporter | @keith      |
   |Type              | Bug    |

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164964440-971a91d6-7abf-4c37-86c8-3e03fd2a3869.gz)

      
<details>
  <summary>Environment</summary>

Xcode 13.2.1 / Swift 5.5

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       |     |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 589ca870689fb9763ae8983b534cd425

  </details>

**Issue Description:**

When you have a xcframework that is vendoring a relocatable object file, when you build a multi-arch binary with dependencies with \`swift build --arch arm64 --arch x86_64\`, you get duplicate symbol errors. This seems to be because XCBuild creates more relocatable object files referencing the xcframework, which bundles one set of the symbols, then the final link attempts to reference the original xcframework and the nested library.

I've attached a repro case for an example. If you run \`swift build\`, it builds fine, but when you run \`swift build --arch arm64 --arch x86_64\` it fails. To repro run: \`./buildlib.sh && swift build --arch arm64 --arch x86_64\`

**Comment by neonichu** [comment]
Works with `xcodebuild`, so I think this might be a consequence of SwiftPM's XCBuild support not using `ld -r`?

**Comment by neonichu** [comment]
Actually, it appears `ld -r` is being used in both cases.

**Comment by neonichu** [comment]
Not sure what the difference is, I cannot spot anything meaningful when comparing the linker commandlines for `lib.o` between `xcodebuild` and `swift build`.

**Comment by keith** [comment]
The difference is that `lib.o` in the XCBuild case is created with this:

```
/Applications/Xcode-13.4.0-RC1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target x86_64-apple-macos10.10 -r -isysroot /Applications/Xcode-13.4.0-RC1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -L/Users/ksmiley/Downloads/SR-15802/xcfwrepro/.build/apple/Products/Debug -L/Applications/Xcode-13.4.0-RC1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib -L/Applications/Xcode-13.4.0-RC1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -F/Users/ksmiley/Downloads/SR-15802/xcfwrepro/.build/apple/Products/Debug -F/Applications/Xcode-13.4.0-RC1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -iframework /Applications/Xcode-13.4.0-RC1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -filelist /Users/ksmiley/Downloads/SR-15802/xcfwrepro/.build/apple/Intermediates.noindex/xcfwrepro.build/Debug/lib.build/Objects-normal/x86_64/lib.o.LinkFileList -nostdlib -Xlinker -object_path_lto -Xlinker /Users/ksmiley/Downloads/SR-15802/xcfwrepro/.build/apple/Intermediates.noindex/xcfwrepro.build/Debug/lib.build/Objects-normal/x86_64/lib.o_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -fobjc-link-runtime -L/Applications/Xcode-13.4.0-RC1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx -L/usr/lib/swift -Xlinker -add_ast_path -Xlinker /Users/ksmiley/Downloads/SR-15802/xcfwrepro/.build/apple/Intermediates.noindex/xcfwrepro.build/Debug/lib.build/Objects-normal/x86_64/lib.swiftmodule -lfoo -Xlinker -dependency_info -Xlinker /Users/ksmiley/Downloads/SR-15802/xcfwrepro/.build/apple/Intermediates.noindex/xcfwrepro.build/Debug/lib.build/Objects-normal/x86_64/lib.o_dependency_info.dat -o /Users/ksmiley/Downloads/SR-15802/xcfwrepro/.build/apple/Intermediates.noindex/xcfwrepro.build/Debug/lib.build/Objects-normal/x86_64/Binary/lib.o
```

which bundles the entire static lib in this object file because it's passing `-lfoo` and the source references it, where the swiftpm case doesn't do this so the `lib.o` from swiftpm correctly only has undefined references to the lib

**Comment by neonichu** [comment]
Right, I am not talking about the difference between the two SwiftPM modes, but about the difference between `xcodebuild` (which work) and `swift build --arch arm64 --arch x86_64` (which doesn't).
